File: checker/checker.pas
{
ID: ndchiph1
PROG: checker
LANG: PASCAL
}

uses math;
const	MAX = 13;
var
	fi,fo: text;
    d: array[1..MAX] of integer;
    a: array[1..MAX] of boolean;
    b: array[2..2*MAX] of boolean;
    c: array[1-MAX..MAX-1] of boolean;
    n,res: longint;

procedure print;
var    	i: integer;
		first: boolean;
begin
	inc(res);
	if (res > 3) then exit;

    //
    for i:= 1 to n-1 do write(fo,d[i],' ');
    write(fo,d[n]);
    writeln(fo);
end;

procedure attempt(i: integer);
var 	j: integer;
begin
	for j:= 1 to n do
    	if a[j] and b[i+j] and c[i-j] then
        begin
        	d[i]:= j;
            if i = n then print
            else
            begin
            	a[j]:= false; b[i+j]:= false; c[i-j]:= false;
                attempt(i+1);
                a[j]:= true; b[i+j]:= true; c[i-j]:= true;
            end;
        end;
end;

procedure process;
begin
	fillchar(a,sizeof(a),true);
    fillchar(b,sizeof(b),true);
    fillchar(c,sizeof(c),true);

    //
    res:= 0;
    attempt(1);
	
    writeln(fo,res);
end;


begin
	assign(fi,'checker.in'); reset(fi);
    assign(fo,'checker.out'); rewrite(fo);
    //
    readln(fi,n);
    process;
    //
    close(fi);
    close(fo);
end.

File: checker/checker.cpp
/*
ID: ndchiph1
PROG: checker
LANG: C++
*/

#include <stdio.h>
#include <memory.h>
#include <stdlib.h>

#define MAX 13
using namespace std;

int n, res;
bool a[MAX+1], b[2*MAX+1], c[2*MAX];
int d[MAX+1];

void init() {
    memset(a,true,sizeof(a));
    memset(b,true,sizeof(b));
    memset(c,true,sizeof(c));
}

void attempt2(int i) {
    int t;
    if (i == 1) t = n/2;
    else t = n;
    
    for (int j=1; j<= t; j++) {
        if (a[j] && b[i+j] && c[i-j+n]) {
            d[i] = j;
            if (i == n) res++;
            else {
                a[j] = b[i+j] = c[i-j+n] = false;
                attempt2(i+1);
                a[j] = b[i+j] = c[i-j+n] = true;
            }
        }
    }
}

void process() {
    init();
    //
    res = 0;
    attempt2(1);
    res *= 2;
    
    if (n % 2 == 1) {
        int j = n/2 + 1;
        a[j] = b[1+j] = c[1-j+n] = false;
        attempt2(2);
    }
    printf("%d\n", res);
}

void printres() {
    res++;
    for (int i=1; i<n; i++) printf("%d ", d[i]);
    printf("%d\n", d[n]);
    
    if (res == 3) {
        process();
        exit(0);
    }
}

void attempt1(int i) {
    for (int j=1; j<= n; j++) {
        if (a[j] && b[i+j] && c[i-j+n]) {
            d[i] = j;
            if (i == n) printres();
            else {
                a[j] = b[i+j] = c[i-j+n] = false;
                attempt1(i+1);
                a[j] = b[i+j] = c[i-j+n] = true;
            }
        }
    }
}

main() {
    freopen("checker.in", "r", stdin);
    freopen("checker.out", "w", stdout);
    
    scanf("%d", &n);
    init();
    attempt1(1);
        
    return 0;
}
